elephant/spike_train_surrogates.py

```diff
@@ -186,54 +218,40 @@ def dither_spikes(spiketrain, dither, n_surrogates=1, decimals=None,
         [0.0 ms, 1000.0 ms])>]
 
     """
+    # The trivial case
     if len(spiketrain) == 0:
-        # return the empty spiketrain n times
+        # return the empty spiketrain n_surrogates times
         return [spiketrain.copy() for _ in range(n_surrogates)]
 
+    # Handle units
     units = spiketrain.units
-    t_start = spiketrain.t_start.rescale(units).magnitude
-    t_stop = spiketrain.t_stop.rescale(units).magnitude
-
-    if refractory_period is None or refractory_period == 0:
-        # Main: generate the surrogates
-        dither = dither.rescale(units).magnitude
-        dithered_spiketrains = \
-            spiketrain.magnitude.reshape((1, len(spiketrain))) \
-            + 2 * dither * np.random.random_sample(
-                (n_surrogates, len(spiketrain))) - dither
-        dithered_spiketrains.sort(axis=1)
-
-        if edges:
-            # Leave out all spikes outside
-            # [spiketrain.t_start, spiketrain.t_stop]
-            dithered_spiketrains = \
-                [train[
-                    np.all([t_start < train, train < t_stop], axis=0)]
-                 for train in dithered_spiketrains]
-        else:
-            # Move all spikes outside
-            # [spiketrain.t_start, spiketrain.t_stop] to the range's ends
-            dithered_spiketrains = np.minimum(
-                np.maximum(dithered_spiketrains, t_start),
-                t_stop)
-
-        dithered_spiketrains = dithered_spiketrains * units
+    dither = dither.rescale(units).magnitude.item()
 
+    if not refractory_period:
+        dithered_spiketrains = _dither_spikes(
+            spiketrain, dither, n_surrogates, edges)
     elif isinstance(refractory_period, pq.Quantity):
+        refractory_period = refractory_period.rescale(units).magnitude.item()
+
         dithered_spiketrains = _dither_spikes_with_refractory_period(
             spiketrain, dither, n_surrogates, refractory_period)
     else:
         raise ValueError("refractory_period must be of type pq.Quantity")
 
     # Round the surrogate data to decimal position, if requested
-    if decimals is not None:
-        dithered_spiketrains = \
-            dithered_spiketrains.rescale(pq.ms).round(decimals).rescale(units)
-
-    # Return the surrogates as list of neo.SpikeTrain
-    return [neo.SpikeTrain(train, t_start=t_start, t_stop=t_stop,
-                           sampling_rate=spiketrain.sampling_rate)
-            for train in dithered_spiketrains]
+    if decimals:
+        return [neo.SpikeTrain(
+                train * units.rescale(pq.ms).round(decimals).rescale(units),
```

Comment:
`rescale` will act on the units here, thanks to Cristiano for pointing out this Issue.